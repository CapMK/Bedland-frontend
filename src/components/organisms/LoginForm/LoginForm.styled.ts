import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { PaletteType } from 'theme/theme.types';

export const LoginFormContainer = styled.form`
  align-items: center;
  display: flex;
  flex-direction: column;
  gap: 1rem;
  padding: 0 2.84375rem 0rem 2.84375rem;
  width: 25.625rem;
`;

export const ResetPasswordLink = styled(Link)<{ palette: PaletteType }>`
  color: ${({ palette }) => (palette as PaletteType).orangeDark};
  display: flex;
  justify-content: center;
`;

export default LoginFormContainer;
