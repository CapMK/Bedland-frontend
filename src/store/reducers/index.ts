import { userReducer } from './user/user_slice';
import { residentReducer } from './resident/resident_slice';
import { buildingReducer } from './building/building_slice';
import { flatReducer } from './flat/flat_slice';
import { votingsReducer } from './votings/votings_slice';
import { paymentReducer } from './payments/payments_slice';

export {
  buildingReducer,
  residentReducer,
  userReducer,
  flatReducer,
  votingsReducer,
  paymentReducer,
};
