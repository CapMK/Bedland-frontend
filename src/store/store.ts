import { combineReducers, configureStore, type PreloadedState } from '@reduxjs/toolkit';
import { CurriedGetDefaultMiddleware } from '@reduxjs/toolkit/dist/getDefaultMiddleware';
import loggerMiddleware from 'redux-logger';
import { userApi } from 'services/user/userApi';
import { votingApi } from 'services/voting/votingApi';
import { buildingsApi } from 'services/buildings/buildingsApi';
import { residentApi } from 'services/resident/residentApi';
import { paymentsApi } from 'services/payments/paymentsApi';
import { flatApi } from 'services/flat/flatApi';
import { managerApi } from 'services/manager/managerApi';

import {
  residentReducer,
  buildingReducer,
  userReducer,
  flatReducer,
  paymentReducer,
  votingsReducer,
} from './reducers/index';

import { managerReducer } from './reducers/manager/manager_slice';

export const rootReducer = combineReducers({
  user: userReducer,
  resident: residentReducer,
  manager: managerReducer,
  building: buildingReducer,
  flat: flatReducer,
  payment: paymentReducer,
  voting: votingsReducer,
  [userApi.reducerPath]: userApi.reducer,
  [votingApi.reducerPath]: votingApi.reducer,
  [votingApi.reducerPath]: votingApi.reducer,
  [flatApi.reducerPath]: flatApi.reducer,
  [managerApi.reducerPath]: managerApi.reducer,
  [residentApi.reducerPath]: residentApi.reducer,
  [buildingsApi.reducerPath]: buildingsApi.reducer,
  [paymentsApi.reducerPath]: paymentsApi.reducer,
});

// Adding the api middleware enables caching, invalidation, polling,
// and other useful features of `rtk-query`.
export const rootMiddleware = (getDefaultMiddleware: CurriedGetDefaultMiddleware) =>
  getDefaultMiddleware().concat(
    loggerMiddleware,
    userApi.middleware,
    buildingsApi.middleware,
    flatApi.middleware,
    managerApi.middleware,
    votingApi.middleware,
    votingApi.middleware,
    residentApi.middleware,
    buildingsApi.middleware,
    paymentsApi.middleware,
  );

export type RootState = ReturnType<typeof rootReducer>;

const storeInit = (preloadedState?: PreloadedState<RootState>) =>
  configureStore({
    reducer: rootReducer,
    middleware: rootMiddleware,
    preloadedState,
  });

export const reduxStore = storeInit();

export type AppStore = ReturnType<typeof storeInit>;
export type AppDispatch = AppStore['dispatch'];
