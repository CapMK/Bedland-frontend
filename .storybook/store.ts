import { enhancer as withReduxEnhancer } from 'addon-redux';
import { combineReducers, configureStore, type PreloadedState } from '@reduxjs/toolkit';
import { CurriedGetDefaultMiddleware } from '@reduxjs/toolkit/dist/getDefaultMiddleware';
import loggerMiddleware from 'redux-logger';
import { userApi } from '../src/services/user/userApi';
import { residentReducer, buildingReducer, userReducer } from '../src/store/reducers/index';

export const rootReducer = combineReducers({
  user: userReducer,
  resident: residentReducer,
  building: buildingReducer,
  [userApi.reducerPath]: userApi.reducer,
});

export const rootMiddleware = (getDefaultMiddleware: CurriedGetDefaultMiddleware) =>
  getDefaultMiddleware().concat(loggerMiddleware, userApi.middleware);

const createEnhancer = () => {
  const enhancers = [];
  if (process.env.NODE_ENV !== 'production') {
    enhancers.push(withReduxEnhancer);
  }
  return enhancers;
};

export type RootState = ReturnType<typeof rootReducer>;

const storeInit = (preloadedState?: PreloadedState<RootState>) =>
  configureStore({
    reducer: rootReducer,
    middleware: rootMiddleware,
    preloadedState,
    enhancers: (defaultEnhancers) => [...createEnhancer(), ...defaultEnhancers],
  });

export const reduxStore = storeInit();
