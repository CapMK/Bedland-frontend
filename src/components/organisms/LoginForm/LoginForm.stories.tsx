import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import ThemeProvider from 'theme/ThemeContext';
import GlobalStyle from 'theme/globalStyles';
import { Provider } from 'react-redux';
import { reduxStore } from 'store/store';
import LoginForm from './LoginForm';

const meta = {
  title: 'Organisms/LoginForm',
  component: LoginForm,
  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <ThemeProvider>
        <Provider store={reduxStore}>
          <GlobalStyle />
          <BrowserRouter>{Story()}</BrowserRouter>
        </Provider>
      </ThemeProvider>
    ),
  ],
} satisfies Meta<typeof LoginForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
