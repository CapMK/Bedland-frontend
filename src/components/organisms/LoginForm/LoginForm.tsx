import { useForm, SubmitHandler, FieldValues } from 'react-hook-form';
import { useContext, useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import routes from 'App/routing/routes';
import ErrorMessage from 'components/atoms/ErrorMessage/ErrorMessage';
import InputWithLabel from 'components/molecules/InputWithLabel/InputWithLabel';
import Button from 'components/atoms/Button/Button';
import * as Styled from 'components/atoms/ErrorMessage/ErrorMessage.styled';
import { ThemeContext } from 'theme/ThemeContext';
import Typography from 'components/atoms/Typography/Typography';
import loginRegexPattern from 'utils/loginRegexPatterns';
import { LOGIN_INPUT_WIDTH } from 'utils/constans';
import ERROR_MESSAGES from 'utils/messages';
import { useGetUserMutation, useLoginMutation } from 'services/user/userApi';
import { LoginRequest } from 'services/user/userApi.types';
import { LoginFormContainer, ResetPasswordLink } from './LoginForm.styled';

const LoginForm = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [errorNoUser, setErrorNoUser] = useState<boolean>(false);
  const { palette } = useContext(ThemeContext);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();

  const [getUser] = useGetUserMutation();
  const [login] = useLoginMutation();

  const onSubmit: SubmitHandler<FieldValues> = async (formData: FieldValues) => {
    const currentUser: LoginRequest = {
      login: formData.login,
      password: formData.password,
    };
    const loginData = await login(currentUser);
    if ('data' in loginData) {
      const data = await getUser({ userId: loginData.data.userId, role: loginData.data.role });
      if ('data' in data) {
        navigate(routes.homeForLoggedIn);
      } else {
        setErrorNoUser(true);
      }
    } else {
      setErrorNoUser(true);
    }
  };

  return (
    <LoginFormContainer onSubmit={handleSubmit(onSubmit)}>
      <Typography variant='header1' color='light' style={{ textAlign: 'center' }}>
        Login
      </Typography>
      <InputWithLabel
        label='Login'
        input='login'
        placeholder='Enter your login'
        type='text'
        register={register}
        regexPattern={loginRegexPattern.login}
      />

      <InputWithLabel
        label='Password'
        input='password'
        placeholder='Enter your password'
        type='password'
        register={register}
        regexPattern={loginRegexPattern.password}
      />
      <Styled.ErrorMsg>
        {errors.password || errors.login || errorNoUser ? (
          <ErrorMessage>{ERROR_MESSAGES.loginPage.errorMessage}</ErrorMessage>
        ) : null}
      </Styled.ErrorMsg>
      <Button wrapper width={LOGIN_INPUT_WIDTH} data-testid='login-button'>
        <Button type='submit' variant='primary'>
          Login
        </Button>
      </Button>

      <ResetPasswordLink palette={palette} to={routes.resetPassword}>
        Forgot your password?
      </ResetPasswordLink>
    </LoginFormContainer>
  );
};

export default LoginForm;
