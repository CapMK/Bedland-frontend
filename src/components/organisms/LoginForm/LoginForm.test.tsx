import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import ThemeProvider from 'theme/ThemeContext';
import { Provider } from 'react-redux';
import { reduxStore } from 'store/store';
import LoginForm from './LoginForm';

describe('Login Form', () => {
  afterEach(cleanup);
  it('renders correctly', () => {
    const snapshot = renderer
      .create(
        <ThemeProvider>
          <BrowserRouter>
            <Provider store={reduxStore}>
              <LoginForm />
            </Provider>
          </BrowserRouter>
        </ThemeProvider>,
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
