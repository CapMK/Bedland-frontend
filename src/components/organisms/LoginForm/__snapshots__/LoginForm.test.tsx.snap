// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Form > renders correctly 1`] = `
.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  width: 100%;
}

.c3 {
  border: 1px solid #919AFF;
  border-radius: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  padding: 0.625rem 1rem 0.625rem 1rem;
  width: 19.875rem;
}

.c3:disabled,
.c3:hover,
.c3:active,
.c3:focus {
  outline: none;
}

.c2 {
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.6875rem;
  width: 19.875rem;
}

.c5 {
  width: 19.875rem;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F56513;
  border: 1px solid #F56513;
  border-radius: 0.5rem;
  color: #FCFEFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5625rem 1rem;
  white-space: nowrap;
  width: 100%;
  font-family: Source Sans Pro,sans-serif;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.4;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c6:hover {
  background-color: #FF7C33;
  border: 1px solid #FF7C33;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
  padding: 0 2.84375rem 0rem 2.84375rem;
  width: 25.625rem;
}

.c7 {
  color: #F56513;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <h1
    style={
      {
        "color": "#FCFEFF",
        "fontFamily": "Poppins, sans-serif",
        "fontSize": "2rem",
        "fontWeight": 600,
        "letterSpacing": "2px",
        "lineHeight": 1.2,
        "textAlign": "center",
      }
    }
  >
    Login
  </h1>
  <div
    className="c1"
  >
    <label
      className="c2"
      htmlFor="login"
      size="medium"
    >
      Login
    </label>
    <input
      className="c3"
      data-testid="login"
      id="login"
      name="login"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter your login"
      type="text"
    />
  </div>
  <div
    className="c1"
  >
    <label
      className="c2"
      htmlFor="password"
      size="medium"
    >
      Password
    </label>
    <input
      className="c3"
      data-testid="password"
      id="password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter your password"
      type="password"
    />
  </div>
  <span
    className="c4"
  />
  <div
    className="c5"
  >
    <button
      className="c6"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="submit"
    >
      Login
    </button>
  </div>
  <a
    className="c7"
    href="/resetPassword"
    onClick={[Function]}
    palette={
      {
        "blueDark": "#4553DD",
        "blueLight": "#6573FF",
        "footer": "#3F4476",
        "greenDark": "#30BA57",
        "greenLight": "#83E57A",
        "greyDark": "#767676",
        "greyLight": "#F5F6F8",
        "inputGrey": "#4B4B4B",
        "light": "#FCFEFF",
        "orangeDark": "#F56513",
        "orangeLight": "#FF7C33",
        "redDark": "#CF0000",
        "redLight": "#FF4438",
        "sidebar": "#0A1122",
        "sidebarTabs": "#CACACA",
        "sidebarTabsPressed": "#919AFF",
        "text": "#1B1A1A",
        "white": "#fff",
        "widgetsShadows": "#D9D9D9",
        "yellow": "#FFD541",
      }
    }
  >
    Forgot your password?
  </a>
</form>
`;
